src/medley/core.cljc

```diff
@@ -467,6 +467,51 @@
              (cons run (part (lazy-seq (drop (count run) s))))))))
       coll))))
 
+(defn partition-with
+  "Applies pred to successive values in coll, splitting it each time `(pred
+  prev-item item)` returns logical true. Returns a lazy seq of partitions.
+  Returns a stateful transducer when no collection is provided."
+  {:added "1.7.0"}
+  ([pred]
+   (fn [rf]
+     (let [part #?(:clj (java.util.ArrayList.) :cljs (array-list))
+           pv (volatile! ::none)]
+       (fn
+         ([] (rf))
+         ([result]
+          (let [result (if (.isEmpty part)
+                         result
+                         (let [v (vec (.toArray part))]
+                           (.clear part)
+                           (unreduced (rf result v))))]
+            (rf result)))
```

Comment:
This could be written:

```clojure
(rf (if (.isEmpty part)
      result
      (let [v (vec (.toArray part))]
        (.clear part)
        (unreduced (rf result v)))))